Use mypy for checking type annotations

Type hinting is extremely useful. This should be enforced from now on to improve the code base.
https://breadcrumbscollector.tech/mypy-how-to-use-it-in-my-project/

For a good documentation about type annotation see:
https://mypy.readthedocs.io/en/latest/getting_started.html

See current issues with
```
tox -e mypy
```


```
setup.py:3: error: Skipping analyzing 'setuptools': found module but no type hints or library stubs
src/sbmlutils/report/sbmlfilters.py:8: error: Need type annotation for 'filters' (hint: "filters: List[<type>] = ...")
src/sbmlutils/converters/mathml.py:11: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/converters/copasi.py:4: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/utils.py:8: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/utils.py:9: error: Skipping analyzing 'depinfo': found module but no type hints or library stubs
src/sbmlutils/report/mathml.py:22: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/report/mathml.py:23: error: Cannot find implementation or library stub for module named 'sympy'
src/sbmlutils/report/mathml.py:24: error: Cannot find implementation or library stub for module named 'sympy.printing.latex'
src/sbmlutils/report/mathml.py:25: error: Cannot find implementation or library stub for module named 'sympy.printing.mathml'
src/sbmlutils/report/mathml.py:156: error: Need type annotation for 'piece_chars' (hint: "piece_chars: List[<type>] = ...")
src/sbmlutils/fbc/fbc.py:5: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/fbc/cobra.py:6: error: Skipping analyzing 'pandas': found module but no type hints or library stubs
src/sbmlutils/fbc/cobra.py:10: error: Cannot find implementation or library stub for module named 'cobra'
src/sbmlutils/distrib/distrib_packages.py:7: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/converters/xpp_helpers.py:8: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/validation.py:6: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/history.py:4: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/metadata/parse_registry.py:7: error: Skipping analyzing 'xmlschema': found module but no type hints or library stubs
src/sbmlutils/io/sbml.py:6: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/io/sbml.py:51: error: Name 'doc' already defined on line 42
src/sbmlutils/io/sbml.py:55: error: Name 'doc' already defined on line 42
src/sbmlutils/distrib/distrib_examples.py:4: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/test/io/test_sbml.py:1: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/test/examples/test_galactose.py:1: error: Skipping analyzing 'roadrunner': found module but no type hints or library stubs
src/sbmlutils/test/examples/test_demo.py:2: error: Skipping analyzing 'roadrunner': found module but no type hints or library stubs
src/sbmlutils/metadata/annotator.py:20: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/metadata/annotator.py:21: error: Skipping analyzing 'pandas': found module but no type hints or library stubs
src/sbmlutils/metadata/annotator.py:57: error: Name 'doc' already defined on line 49
src/sbmlutils/metadata/annotator.py:442: error: "ExternalAnnotation" has no attribute "annotation_type"; maybe "_annotation_types"?
src/sbmlutils/metadata/annotator.py:444: error: "ExternalAnnotation" has no attribute "resource"
src/sbmlutils/metadata/annotator.py:452: error: "ExternalAnnotation" has no attribute "annotation_type"; maybe "_annotation_types"?
src/sbmlutils/metadata/annotator.py:454: error: "ExternalAnnotation" has no attribute "sbml_type"; maybe "_sbml_types"?
src/sbmlutils/metadata/annotator.py:466: error: "ExternalAnnotation" has no attribute "annotation_type"; maybe "_annotation_types"?
src/sbmlutils/metadata/annotator.py:467: error: "ExternalAnnotation" has no attribute "resource"
src/sbmlutils/metadata/annotator.py:468: error: "ExternalAnnotation" has no attribute "annotation_type"; maybe "_annotation_types"?
src/sbmlutils/metadata/annotator.py:469: error: "ExternalAnnotation" has no attribute "resource"
src/sbmlutils/metadata/annotator.py:472: error: "ExternalAnnotation" has no attribute "annotation_type"; maybe "_annotation_types"?
src/sbmlutils/metadata/annotator.py:565: error: Bracketed expression "[...]" is not valid as a type
src/sbmlutils/metadata/annotator.py:565: note: Did you mean "List[...]"?
src/sbmlutils/manipulation/interpolation.py:9: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/manipulation/interpolation.py:10: error: Skipping analyzing 'pandas': found module but no type hints or library stubs
src/sbmlutils/converters/odefac.py:28: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/report/formating.py:2: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/report/formating.py:116: error: Incompatible return value type (got "None", expected "str")
src/sbmlutils/comp/flatten.py:7: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/report/sbmlreport.py:22: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/report/sbmlreport.py:76: error: Argument 1 to "_create_index_html" has incompatible type "Iterable[Path]"; expected "List[Path]"
src/sbmlutils/report/sbmlreport.py:347: error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
src/sbmlutils/report/sbmlreport.py:349: error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
src/sbmlutils/report/sbmlreport.py:426: error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
src/sbmlutils/report/sbmlreport.py:428: error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
src/sbmlutils/report/sbmlreport.py:743: error: Incompatible types in assignment (expression has type "str", variable has type "List[Any]")
src/sbmlutils/report/sbmlreport.py:745: error: Incompatible types in assignment (expression has type "str", variable has type "List[Any]")
src/sbmlutils/distrib/distrib_report.py:15: error: Unexpected keyword argument "target_dir" for "create_report"
src/sbmlutils/report/sbmlreport.py:82: note: "create_report" defined here
src/sbmlutils/distrib/distrib_report.py:15: error: Argument 1 to "create_report" has incompatible type "str"; expected "Path"
src/sbmlutils/factory.py:23: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/factory.py:546: error: Signature of "set_fields" incompatible with supertype "Sbase"
src/sbmlutils/factory.py:1081: error: Name 'up' already defined on line 1073
src/sbmlutils/factory.py:1082: error: "None" has no attribute "setType"
src/sbmlutils/factory.py:1084: error: "None" has no attribute "setValueLower"
src/sbmlutils/factory.py:1086: error: "None" has no attribute "setValueUpper"
src/sbmlutils/factory.py:1088: error: "None" has no attribute "setVarLower"
src/sbmlutils/factory.py:1090: error: "None" has no attribute "setValueLower"
src/sbmlutils/factory.py:1104: error: Name 'up' already defined on line 1073
src/sbmlutils/factory.py:1107: error: "None" has no attribute "setType"
src/sbmlutils/factory.py:1109: error: "None" has no attribute "setValue"
src/sbmlutils/factory.py:1111: error: "None" has no attribute "setValue"
src/sbmlutils/factory.py:1124: error: Name 'up' already defined on line 1073
src/sbmlutils/factory.py:1125: error: "None" has no attribute "setType"
src/sbmlutils/factory.py:1141: error: "None" has no attribute "setDefinitionURL"
src/sbmlutils/factory.py:1148: error: "None" has no attribute "setMath"
src/sbmlutils/factory.py:1343: error: Signature of "set_fields" incompatible with supertype "Sbase"
src/sbmlutils/factory.py:1416: error: Signature of "set_fields" incompatible with supertype "Sbase"
src/sbmlutils/test/report/test_mathml.py:1: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/test/manipulation/test_interpolation.py:8: error: Skipping analyzing 'pandas': found module but no type hints or library stubs
src/sbmlutils/test/manipulation/test_interpolation.py:10: error: Skipping analyzing 'roadrunner': found module but no type hints or library stubs
src/sbmlutils/units.py:3: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/creator.py:20: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/creator.py:21: error: Skipping analyzing 'xmltodict': found module but no type hints or library stubs
src/sbmlutils/creator.py:46: error: Need type annotation for 'cdict' (hint: "cdict: Dict[<type>, <type>] = ...")
src/sbmlutils/creator.py:164: error: "CoreModel" has no attribute "version"
src/sbmlutils/creator.py:165: error: "CoreModel" has no attribute "mid"
src/sbmlutils/creator.py:165: error: "CoreModel" has no attribute "version"
src/sbmlutils/creator.py:167: error: "CoreModel" has no attribute "mid"
src/sbmlutils/creator.py:185: error: Access to generic instance variables via class is ambiguous
src/sbmlutils/creator.py:189: error: Access to generic instance variables via class is ambiguous
src/sbmlutils/creator.py:235: error: "CoreModel" has no attribute "packages"
src/sbmlutils/creator.py:249: error: "CoreModel" has no attribute "packages"
src/sbmlutils/creator.py:253: error: "CoreModel" has no attribute "packages"
src/sbmlutils/creator.py:263: error: "CoreModel" has no attribute "notes"
src/sbmlutils/creator.py:264: error: "CoreModel" has no attribute "notes"
src/sbmlutils/creator.py:267: error: "CoreModel" has no attribute "creators"
src/sbmlutils/creator.py:271: error: "CoreModel" has no attribute "model_units"
src/sbmlutils/creator.py:382: error: Argument 1 to "dict_from_modules" of "Preprocess" has incompatible type "Union[Iterable[str], Dict[Any, Any]]"; expected "List[str]"
src/sbmlutils/creator.py:397: error: Incompatible types in assignment (expression has type "str", variable has type "Optional[Path]")
src/sbmlutils/creator.py:398: error: Argument 1 to "join" has incompatible type "Optional[Path]"; expected "Union[str, _PathLike[str]]"
src/sbmlutils/creator.py:404: error: Item "None" of "Optional[Path]" has no attribute "exists"
src/sbmlutils/creator.py:406: error: Item "None" of "Optional[Path]" has no attribute "mkdir"
src/sbmlutils/creator.py:407: error: Unsupported left operand type for / ("None")
src/sbmlutils/creator.py:407: note: Left operand is of type "Optional[Path]"
src/sbmlutils/creator.py:407: error: Incompatible types in assignment (expression has type "Union[Path, Any]", variable has type "str")
src/sbmlutils/creator.py:415: error: Argument "filepath" to "write_sbml" has incompatible type "str"; expected "Optional[Path]"
src/sbmlutils/creator.py:427: error: Argument "filepath" to "annotate_sbml" has incompatible type "str"; expected "Path"
src/sbmlutils/creator.py:434: error: Argument "sbml_path" to "create_report" has incompatible type "str"; expected "Path"
src/sbmlutils/creator.py:434: error: Argument "output_dir" to "create_report" has incompatible type "Optional[Path]"; expected "Path"
src/sbmlutils/creator.py:438: error: Argument 1 to "rmtree" has incompatible type "Optional[Path]"; expected "Union[bytes, Union[str, _PathLike[str]]]"
src/sbmlutils/creator.py:443: error: Argument "sbml_path" to "FactoryResult" has incompatible type "str"; expected "Path"
src/sbmlutils/test/test_history.py:4: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/layout/layout.py:3: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/layout/layout.py:75: error: Name 'layout_model' already defined on line 64
src/sbmlutils/layout/layout.py:132: error: Signature of "set_fields" incompatible with supertype "Sbase"
src/sbmlutils/layout/layout.py:191: error: Signature of "set_fields" incompatible with supertype "Sbase"
src/sbmlutils/layout/layout.py:209: error: Unexpected keyword argument "dpeth" for "_create_bounding_box"; did you mean "depth"?
src/sbmlutils/layout/layout.py:255: error: Signature of "set_fields" incompatible with supertype "Sbase"
src/sbmlutils/converters/xpp.py:47: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/converters/xpp.py:217: error: Need type annotation for 'function_definitions' (hint: "function_definitions: List[<type>] = ...")
src/sbmlutils/converters/xpp.py:218: error: Need type annotation for 'events' (hint: "events: List[<type>] = ...")
src/sbmlutils/comp/comp.py:13: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/comp/comp.py:13: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
src/sbmlutils/test/test_formating.py:1: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/test/test_factory.py:2: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/test/test_distrib.py:3: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/examples/fbc2.py:218: error: "Reaction" has no attribute "flux_unit"
src/sbmlutils/examples/fbc1.py:211: error: "Reaction" has no attribute "flux_unit"
src/sbmlutils/examples/distrib_uncertainty.py:2: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/examples/core2.py:47: error: Need type annotation for 'functions' (hint: "functions: List[<type>] = ...")
src/sbmlutils/examples/core2.py:78: error: Need type annotation for 'rules' (hint: "rules: List[<type>] = ...")
src/sbmlutils/examples/core2.py:79: error: Need type annotation for 'reactions' (hint: "reactions: List[<type>] = ...")
src/sbmlutils/examples/dallaman/model.py:142: error: Need type annotation for 'functions' (hint: "functions: List[<type>] = ...")
src/sbmlutils/examples/dallaman/model.py:147: error: Need type annotation for 'compartments' (hint: "compartments: List[<type>] = ...")
src/sbmlutils/examples/dallaman/model.py:152: error: Need type annotation for 'species' (hint: "species: List[<type>] = ...")
src/sbmlutils/examples/dallaman/model.py:238: error: Need type annotation for 'assignments' (hint: "assignments: List[<type>] = ...")
src/sbmlutils/examples/dallaman/model.py:304: error: Need type annotation for 'reactions' (hint: "reactions: List[<type>] = ...")
src/sbmlutils/converters/xpp_examples.py:5: error: Skipping analyzing 'roadrunner': found module but no type hints or library stubs
src/sbmlutils/converters/xpp_examples.py:6: error: Skipping analyzing 'matplotlib': found module but no type hints or library stubs
src/sbmlutils/test/test_comp.py:4: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/test/test_annotation.py:6: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/test/test_annotation.py:51: error: Name 'test_model_annotation' already defined on line 28
src/sbmlutils/manipulation/merge.py:12: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/manipulation/merge.py:44: error: Argument 1 to "chdir" has incompatible type "Optional[Path]"; expected "Union[int, Union[str, bytes, _PathLike[str], _PathLike[bytes]]]"
src/sbmlutils/manipulation/merge.py:63: error: Unsupported left operand type for / ("None")
src/sbmlutils/manipulation/merge.py:63: note: Left operand is of type "Optional[Path]"
src/sbmlutils/manipulation/merge.py:71: error: Incompatible types in assignment (expression has type "str", variable has type "Path")
src/sbmlutils/manipulation/merge.py:72: error: Module has no attribute "check_sbml"
src/sbmlutils/manipulation/merge.py:79: error: Module has no attribute "check_sbml"
src/sbmlutils/manipulation/merge.py:82: error: Argument 1 to "join" has incompatible type "Optional[Path]"; expected "Union[str, _PathLike[str]]"
src/sbmlutils/examples/distrib_comp.py:6: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/examples/tiny_model/model.py:346: error: List item 0 has incompatible type "Constraint"; expected "Event"
src/sbmlutils/examples/tiny_model/simulation.py:12: error: Skipping analyzing 'pandas': found module but no type hints or library stubs
src/sbmlutils/examples/tiny_model/simulation.py:13: error: Skipping analyzing 'roadrunner': found module but no type hints or library stubs
src/sbmlutils/examples/tiny_model/simulation.py:14: error: Skipping analyzing 'matplotlib': found module but no type hints or library stubs
src/sbmlutils/examples/tiny_model/simulation.py:30: error: Incompatible types in assignment (expression has type "CoreModel", variable has type "Dict[Any, Any]")
Found 149 errors in 49 files (checked 100 source files)
```

@sankha555 Please fix the mypy issues for the sbmlreport
```
src/sbmlutils/report/sbmlreport.py:22: error: Skipping analyzing 'libsbml': found module but no type hints or library stubs
src/sbmlutils/report/sbmlreport.py:76: error: Argument 1 to "_create_index_html" has incompatible type "Iterable[Path]"; expected "List[Path]"
src/sbmlutils/report/sbmlreport.py:347: error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
src/sbmlutils/report/sbmlreport.py:349: error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
src/sbmlutils/report/sbmlreport.py:426: error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
src/sbmlutils/report/sbmlreport.py:428: error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
src/sbmlutils/report/sbmlreport.py:743: error: Incompatible types in assignment (expression has type "str", variable has type "List[Any]")
src/sbmlutils/report/sbmlreport.py:745: error: Incompatible types in assignment (expression has type "str", variable has type "List[Any]")
```

--- a/src/sbmlutils/report/sbmlreport.py
+++ b/src/sbmlutils/report/sbmlreport.py
@@ -16,7 +16,7 @@ import logging
 import ntpath
 import warnings
 from pathlib import Path
-from typing import Dict, Iterable, List
+from typing import Dict, List
 
 import jinja2
 import libsbml
@@ -33,7 +33,7 @@ TEMPLATE_DIR = RESOURCES_DIR / "templates"
 
 
 def create_reports(
-    sbml_paths: Iterable[Path],
+    sbml_paths: List[Path],
     output_dir: Path,
     template: str = "report.html",
     promote: bool = False,
@@ -344,10 +344,10 @@ def infoSbase(item: libsbml.SBase) -> Dict:
                     """
                 )
             if len(replaced_elements) == 0:
-                replaced_elements = ""
+                replaced_elements_combined = ""
             else:
-                replaced_elements = "".join(replaced_elements)
-            info["replaced_elements"] = replaced_elements
+                replaced_elements_combined = "".join(replaced_elements)
+            info["replaced_elements"] = replaced_elements_combined
 
     return info
 
@@ -423,10 +423,10 @@ def listOfSubmodels_dict(model: libsbml.Model) -> List[Dict]:
             for deletion in item.getListOfDeletions():
                 deletions.append(sbaseref(deletion))
             if len(deletions) == 0:
-                deletions = empty_html()
+                deletions_combined = empty_html()
             else:
-                deletions = "<br />".join(deletions)
-            info["deletions"] = deletions
+                deletions_combined = "<br />".join(deletions)
+            info["deletions"] = deletions_combined
 
             time_conversion = empty_html()
             if item.isSetTimeConversionFactor():
@@ -740,10 +740,10 @@ def listOfReactions_dict(model: libsbml.Model, math_type: str) -> List[Dict]:
         for mod in item.getListOfModifiers():
             modifiers.append(mod.getSpecies())
         if len(modifiers) == 0:
-            modifiers = empty_html()
+            modifiers_html = empty_html()
         else:
-            modifiers = "<br />".join(modifiers)
-        info["modifiers"] = modifiers
+            modifiers_html = "<br />".join(modifiers)
+        info["modifiers"] = modifiers_html
         klaw = item.getKineticLaw()
         info["formula"] = math(klaw, math_type)
         info["derived_units"] = derived_units(klaw)

